File: src/main/vrml/net/coderextreme/data/bubble.mjs
const
   X3D     = require ("x_ite-node"),
   canvas  = X3D .createBrowser (),
   browser = canvas .browser,
   scene   = browser .currentScene;
async function main () {
scene.setProfile(browser.getProfile("Immersive"));
scene.addMetaData("title", "bubble.x3d");
scene.addMetaData("creator", "John Carlson");
scene.addMetaData("description", "Tour around a prismatic sphere");
scene.addMetaData("generator", "X3D-Edit, https://savage.nps.edu/X3D-Edit");
scene.addMetaData("identifier", "https://coderextreme.net/X3DJSONLD/src/main/data/bubble.x3d");
await browser .loadComponents (scene);
let NavigationInfo8 = browser.currentScene.createNode("NavigationInfo");
NavigationInfo8.type = new X3D.MFString([new X3D.SFString("EXAMINE")]);
browser.currentScene.children = new X3D.MFNode();

browser.currentScene.children[0] = NavigationInfo8;

let Viewpoint9 = browser.currentScene.createNode("Viewpoint");
Viewpoint9.position = new X3D.SFVec3f([0,0,4]);
Viewpoint9.orientation = new X3D.SFRotation([1,0,0,0]);
Viewpoint9.description = "Bubble in action";
browser.currentScene.children[1] = Viewpoint9;

let ProtoDeclare10 = browser.createX3DFromString(`<?xml version="1.0" encoding="undefined"?>
<!DOCTYPE X3D PUBLIC "ISO//Web3D//DTD X3D undefined//EN" "https://www.web3d.org/specifications/x3d-undefined.dtd">
<ProtoDeclare name="Bubble" ><ProtoBody><Transform DEF="transform"><Shape><Sphere radius="0.25"></Sphere>
<Appearance containerField="appearance"><Material containerField="material" diffuseColor="1 0 0" transparency="0.2"></Material>
</Appearance>
</Shape>
<Script DEF="bounce"><field name="scale" accessType="inputOutput" type="SFVec3f" value="1 1 1"></field>
<field name="translation" accessType="inputOutput" type="SFVec3f" value="0 0 0"></field>
<field name="velocity" accessType="inputOutput" type="SFVec3f" value="0 0 0"></field>
<field name="scalvel" accessType="inputOutput" type="SFVec3f" value="0 0 0"></field>
<field name="set_fraction" accessType="inputOnly" type="SFFloat"></field>
<![CDATA[ecmascript:
function initialize() {
    velocity = new SFVec3f(Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125);

    scalvel = new SFVec3f(Math.random() * 0.4, Math.random() * 0.4, Math.random() * 0.4);
}

function set_fraction(value) {
	translation = new SFVec3f(
		translation.x + velocity.x,
		translation.y + velocity.y,
		translation.z + velocity.z);
	scale = new SFVec3f(
		scale.x + scalvel.x,
		scale.y + scalvel.y,
		scale.z + scalvel.z);
        // if you get to far away or too big, explode
        if ( Math.abs(translation.x) > 256) {
		translation.x = 0;
		initialize();
	}
        if ( Math.abs(translation.y) > 256) {
		translation.y = 0;
		initialize();
	}
        if ( Math.abs(translation.z) > 256) {
		translation.z = 0;
		initialize();
	}
	if (Math.abs(scale.x) > 20) {
		scale.x = scale.x/2;
		translation.x = 0;
		initialize();
	}
	if (Math.abs(scale.y) > 20) {
		scale.y = scale.y/2;
		translation.y = 0;
		initialize();
	}
	if (Math.abs(scale.z) > 20) {
		scale.z = scale.z/2;
		translation.z = 0;
		initialize();
	}
}]]></Script>
<TimeSensor DEF="bubbleClock" cycleInterval="10" loop="true"></TimeSensor>
<ROUTE fromNode="bounce" fromField="translation_changed" toNode="transform" toField="set_translation"></ROUTE>
<ROUTE fromNode="bounce" fromField="scale_changed" toNode="transform" toField="set_scale"></ROUTE>
<ROUTE fromNode="bubbleClock" fromField="fraction_changed" toNode="bounce" toField="set_fraction"></ROUTE>
</Transform>
</ProtoBody>
</ProtoDeclare>`);
ProtoDeclare10.name = "Bubble";
let ProtoBody11 = browser.currentScene.createNode("ProtoBody");
let Transform12 = browser.currentScene.createNode("Transform");
Transform12.DEF = "transform";
let Shape13 = browser.currentScene.createNode("Shape");
let Sphere14 = browser.currentScene.createNode("Sphere");
Sphere14.radius = 0.25;
geometry = Sphere14;

let Appearance15 = browser.currentScene.createNode("Appearance");
let Material16 = browser.currentScene.createNode("Material");
Material16.diffuseColor = new X3D.SFColor([1,0,0]);
Material16.transparency = 0.2;
material = Material16;

appearance = Appearance15;

Transform12YYY.child = new X3D.undefined();

Transform12ZZZ.child[0] = Shape13;

let Script17 = browser.currentScene.createNode("Script");
Script17.DEF = "bounce";
Script18.getField("scale").setValue("1 1 1");
Script17YYY.field = new X3D.MFNode();

Script19.getField("translation").setValue("0 0 0");
Script17YYY.field = new X3D.MFNode();

Script20.getField("velocity").setValue("0 0 0");
Script17YYY.field = new X3D.MFNode();

Script21.getField("scalvel").setValue("0 0 0");
Script17YYY.field = new X3D.MFNode();

Script22.getField("set_fraction")Script17YYY.field = new X3D.MFNode();


Script17.setSourceCode(`ecmascript:\n"+
"function initialize() {\n"+
"    velocity = new SFVec3f(Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125);\n"+
"\n"+
"    scalvel = new SFVec3f(Math.random() * 0.4, Math.random() * 0.4, Math.random() * 0.4);\n"+
"}\n"+
"\n"+
"function set_fraction(value) {\n"+
"	translation = new SFVec3f(\n"+
"		translation.x + velocity.x,\n"+
"		translation.y + velocity.y,\n"+
"		translation.z + velocity.z);\n"+
"	scale = new SFVec3f(\n"+
"		scale.x + scalvel.x,\n"+
"		scale.y + scalvel.y,\n"+
"		scale.z + scalvel.z);\n"+
"        // if you get to far away or too big, explode\n"+
"        if ( Math.abs(translation.x) > 256) {\n"+
"		translation.x = 0;\n"+
"		initialize();\n"+
"	}\n"+
"        if ( Math.abs(translation.y) > 256) {\n"+
"		translation.y = 0;\n"+
"		initialize();\n"+
"	}\n"+
"        if ( Math.abs(translation.z) > 256) {\n"+
"		translation.z = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.x) > 20) {\n"+
"		scale.x = scale.x/2;\n"+
"		translation.x = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.y) > 20) {\n"+
"		scale.y = scale.y/2;\n"+
"		translation.y = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.z) > 20) {\n"+
"		scale.z = scale.z/2;\n"+
"		translation.z = 0;\n"+
"		initialize();\n"+
"	}\n"+
"}`)
Transform12ZZZ.children[1] = Script17;

let TimeSensor23 = browser.currentScene.createNode("TimeSensor");
TimeSensor23.DEF = "bubbleClock";
TimeSensor23.cycleInterval = 10;
TimeSensor23.loop = True;
Transform12ZZZ.children[2] = TimeSensor23;

let ROUTE24 = browser.currentScene.createNode("ROUTE");
ROUTE24.fromNode = "bounce";
ROUTE24.fromField = "translation_changed";
ROUTE24.toNode = "transform";
ROUTE24.toField = "set_translation";
Transform12ZZZ.children[3] = ROUTE24;

let ROUTE25 = browser.currentScene.createNode("ROUTE");
ROUTE25.fromNode = "bounce";
ROUTE25.fromField = "scale_changed";
ROUTE25.toNode = "transform";
ROUTE25.toField = "set_scale";
Transform12ZZZ.children[4] = ROUTE25;

let ROUTE26 = browser.currentScene.createNode("ROUTE");
ROUTE26.fromNode = "bubbleClock";
ROUTE26.fromField = "fraction_changed";
ROUTE26.toNode = "bounce";
ROUTE26.toField = "set_fraction";
Transform12ZZZ.children[5] = ROUTE26;

ProtoBody11YYY.children = new X3D.MFNode();

ProtoBody11ZZZ.children[0] = Transform12;

protoBody = ProtoBody11;

browser.currentScene.children[2] = ProtoDeclare10;

let ProtoInstance27 = browser.currentScene.createNode("ProtoInstance");
ProtoInstance27.name = "Bubble";
ProtoInstance27.DEF = "bubbleA";
browser.currentScene.children[3] = ProtoInstance27;

}
main ();

File: src/main/quarter/net/coderextreme/personal/bubble.cpp

#include <Inventor/nodes/SoCone.h>
#include <Inventor/nodes/SoBaseColor.h>
#include <Inventor/nodes/SoSeparator.h>
#include <Inventor/nodes/SoShape.h>
#include <Inventor/nodes/SoTransform.h>
#include <Inventor/nodes/SoSphere.h>
#include <Inventor/nodes/SoMaterial.h>
#include <Inventor/misc/SoProtoInstance.h>
#include <Inventor/VRMLnodes/SoVRMLAppearance.h>
#include <Quarter/Quarter.h>
#include <Quarter/QuarterWidget.h>
#include <QApplication>
#include <QMainWindow>

using namespace SIM::Coin3D::Quarter;
int main(int argc, char ** argv) 
{

  QApplication app(argc, argv);
  Quarter::init();

  SoSeparator * root = new SoSeparator;
  root->ref();

  SoBaseColor * col = new SoBaseColor;
  col->rgb = SbColor(1, 1, 0);
  root->addChild(col);

  root->addChild(new SoCone);

  QMainWindow * mainwin = new QMainWindow();

  QuarterWidget * viewer = new QuarterWidget(mainwin);
  viewer->setNavigationModeFile();
  mainwin->setCentralWidget(viewer);
  viewer->setSceneGraph(root);

  mainwin->show();
  app.exec();
  root->unref();
  delete viewer;

  Quarter::clean();
SoSceneManager* SoSceneManager0 = new SoSceneManager();
SoSceneManager0->setProfile(QString("Immersive"));
SoSceneManager0->setVersion(QString("3.3"));
Sohead* Sohead1 = new Sohead();
Someta* Someta2 = new Someta();
Someta2->setName(QString("title"));
Someta2->setContent(QString("bubble.x3d"));
Sohead1->addMeta(*Someta2);

Someta* Someta3 = new Someta();
Someta3->setName(QString("creator"));
Someta3->setContent(QString("John Carlson"));
Sohead1->addMeta(*Someta3);

Someta* Someta4 = new Someta();
Someta4->setName(QString("description"));
Someta4->setContent(QString("Tour around a prismatic sphere"));
Sohead1->addMeta(*Someta4);

Someta* Someta5 = new Someta();
Someta5->setName(QString("generator"));
Someta5->setContent(QString("X3D-Edit, https://savage.nps.edu/X3D-Edit"));
Sohead1->addMeta(*Someta5);

Someta* Someta6 = new Someta();
Someta6->setName(QString("identifier"));
Someta6->setContent(QString("https://coderextreme.net/X3DJSONLD/src/main/personal/bubble.x3d"));
Sohead1->addMeta(*Someta6);

SoSceneManager0->setHead(*Sohead1);

SoNode* SoNode7 = new SoNode();
SoNavigationInfo* SoNavigationInfo8 = new SoNavigationInfo();
SoNavigationInfo8->setType(new QString[]{QString("EXAMINE")}, 1);
SoNode7->addChild(*SoNavigationInfo8);

SoViewpoint* SoViewpoint9 = new SoViewpoint();
SoViewpoint9->setPosition(new float[]{0.0,0.0,4.0});
SoViewpoint9->setOrientation(new float[]{1.0,0.0,0.0,0.0});
SoViewpoint9->setDescription(QString("Bubble in action"));
SoNode7->addChild(*SoViewpoint9);

SoProtoDeclare* SoProtoDeclare10 = new SoProtoDeclare();
SoProtoDeclare10->setName(QString("Bubble"));
SoProtoBody* SoProtoBody11 = new SoProtoBody();
SoTransform* SoTransform12 = new SoTransform();
SoTransform12->setDEF(QString("transform"));
SoShape* SoShape13 = new SoShape();
SoSphere* SoSphere14 = new SoSphere();
SoSphere14->setRadius(0.25);
SoShape13->setGeometry(*SoSphere14);

SoVRMLAppearance* SoVRMLAppearance15 = new SoVRMLAppearance();
SoMaterial* SoMaterial16 = new SoMaterial();
SoMaterial16->setDiffuseColor(new float[]{1.0,0.0,0.0});
SoMaterial16->setTransparency(0.2);
SoVRMLAppearance15->addChild(*SoMaterial16);

SoShape13->addChild(*SoVRMLAppearance15);

SoTransform12->addChild(*SoShape13);

SoScript* SoScript17 = new SoScript();
SoScript17->setDEF(QString("bounce"));
Sofield* Sofield18 = new Sofield();
Sofield18->setName(QString("scale"));
Sofield18->setAccessType(QString("inputOutput"));
Sofield18->setType(QString("SFVec3f"));
Sofield18->setValue(QString("1 1 1"));
SoScript17->addChild(*Sofield18);

Sofield* Sofield19 = new Sofield();
Sofield19->setName(QString("translation"));
Sofield19->setAccessType(QString("inputOutput"));
Sofield19->setType(QString("SFVec3f"));
Sofield19->setValue(QString("0 0 0"));
SoScript17->addChild(*Sofield19);

Sofield* Sofield20 = new Sofield();
Sofield20->setName(QString("velocity"));
Sofield20->setAccessType(QString("inputOutput"));
Sofield20->setType(QString("SFVec3f"));
Sofield20->setValue(QString("0 0 0"));
SoScript17->addChild(*Sofield20);

Sofield* Sofield21 = new Sofield();
Sofield21->setName(QString("scalvel"));
Sofield21->setAccessType(QString("inputOutput"));
Sofield21->setType(QString("SFVec3f"));
Sofield21->setValue(QString("0 0 0"));
SoScript17->addChild(*Sofield21);

Sofield* Sofield22 = new Sofield();
Sofield22->setName(QString("set_fraction"));
Sofield22->setAccessType(QString("inputOnly"));
Sofield22->setType(QString("SFFloat"));
SoScript17->addChild(*Sofield22);


//SoScript17->setSourceCode(QString("ecmascript:")+
//_T("function initialize() {")+
//_T("    velocity = new SFVec3f(Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125);")+
//_T("    scalvel = new SFVec3f(Math.random() * 0.4, Math.random() * 0.4, Math.random() * 0.4);")+
//_T("}")+
//_T("function set_fraction(value) {")+
//_T("	translation = new SFVec3f(")+
//_T("		translation.x + velocity.x,")+
//_T("		translation.y + velocity.y,")+
//_T("		translation.z + velocity.z);")+
//_T("	scale = new SFVec3f(")+
//_T("		scale.x + scalvel.x,")+
//_T("		scale.y + scalvel.y,")+
//_T("		scale.z + scalvel.z);")+
//_T("        // if you get to far away or too big, explode")+
//_T("        if ( Math.abs(translation.x) > 256) {")+
//_T("		translation.x = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("        if ( Math.abs(translation.y) > 256) {")+
//_T("		translation.y = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("        if ( Math.abs(translation.z) > 256) {")+
//_T("		translation.z = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("	if (Math.abs(scale.x) > 20) {")+
//_T("		scale.x = scale.x/2;")+
//_T("		translation.x = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("	if (Math.abs(scale.y) > 20) {")+
//_T("		scale.y = scale.y/2;")+
//_T("		translation.y = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("	if (Math.abs(scale.z) > 20) {")+
//_T("		scale.z = scale.z/2;")+
//_T("		translation.z = 0;")+
//_T("		initialize();")+
//_T("	}")+
//_T("}"));
SoTransform12->addChild(*SoScript17);

SoTimeSensor* SoTimeSensor23 = new SoTimeSensor();
SoTimeSensor23->setDEF(QString("bubbleClock"));
SoTimeSensor23->setCycleInterval(10);
SoTimeSensor23->setLoop(true);
SoTransform12->addChild(*SoTimeSensor23);

SoROUTE* SoROUTE24 = new SoROUTE();
SoROUTE24->setFromNode(QString("bounce"));
SoROUTE24->setFromField(QString("translation_changed"));
SoROUTE24->setToNode(QString("transform"));
SoROUTE24->setToField(QString("set_translation"));
SoTransform12->addChild(*SoROUTE24);

SoROUTE* SoROUTE25 = new SoROUTE();
SoROUTE25->setFromNode(QString("bounce"));
SoROUTE25->setFromField(QString("scale_changed"));
SoROUTE25->setToNode(QString("transform"));
SoROUTE25->setToField(QString("set_scale"));
SoTransform12->addChild(*SoROUTE25);

SoROUTE* SoROUTE26 = new SoROUTE();
SoROUTE26->setFromNode(QString("bubbleClock"));
SoROUTE26->setFromField(QString("fraction_changed"));
SoROUTE26->setToNode(QString("bounce"));
SoROUTE26->setToField(QString("set_fraction"));
SoTransform12->addChild(*SoROUTE26);

SoProtoBody11->addChild(*SoTransform12);

SoProtoDeclare10->addChild(*SoProtoBody11);

SoNode7->addChild(*SoProtoDeclare10);

SoProtoInstance* SoProtoInstance27 = new SoProtoInstance();
SoProtoInstance27->setName(QString("Bubble"));
SoProtoInstance27->setDEF(QString("bubbleA"));
SoNode7->addChild(*SoProtoInstance27);

SoSceneManager0->setSceneGraph(*SoNode7);

return 0;
}
